### PR TITLE
Feat/card review summary

### LINKED ISSUE #3: feat: create reviews summary for a better ux
### Feature Request

With a number of reviews comes a number of possibilities...

### Definition of Done

- [x] A user must be able to read the number of reviews done to a visited grant program
- [x] A user must be able to read the average review score for a visited grant program, if 1 review or more were done
- [x] A user must be able to read the number of reviews done per score, for a visited grant program, if 1 review or more were done
- [x] A user must be able to read the average interval that a grant program is reviewed, if 1 review or more were done
- [x] A user must be able to click the "+ Add Review" button
- [x] A user must be able to see this nicely displayed in mobile, tablet and desktop views. Please adjust the design as necessary to achieve so (Example: instead of using vertical dividers in this section, as proposed for desktop views, use horizontal ones for mobile and tablet views)

### How to do so?

1. Number of reviews
Simply gather the length of the reviews array of the visited grant program.

2. Average interval
Gather the timestamp of the first review of a grant program and gather the timestamp of the "now" moment. With T1 and T2 (timestamp 1 and 2), do T1 - T2 to get to know the interval time between these times. Divide this interval by the number of reviews the grant program has. This result, this timestamp, represents the interval of time that it is taking for the grant program to receive a review. Lastly, transform the timestamp to a readable information: use "per day", "per week", "per month", "per year" labels if the interval is less than a 1 year. 

Example: 

3 days interval messages: "Typically reviewed 2 times per week" 
1 day interval messages: "Typically reviewed 1 time per day" 
7 days interval messages: "Typically reviewed 1 time per week" 
21 days interval messages: "Typically reviewed 1 time per month" 
18 days interval messages: "Typically reviewed 1 time per month" 
32 days interval messages: "Typically reviewed 11 times per year" 
365 days interval messages: "Typically reviewed 1 time per year" 
366 days interval messages: "Typically reviewed each 366 days" 

3. Average review
Sum all the reviews scores and divide this number by the number of reviews summed.

4. Reviews per score
Math.round() each review and set each one into a bucket between the 5 possible ones (1 star, 2 stars, 3 stars, 4 stars or 5 stars)... After this, divide each bucket length by the total number of reviews to gather the percentage each bucket has of all the reviews.

### Design specifications

Can be found in this [Figma artboard](https://www.figma.com/design/phBIuObI840Cd7sKvZnuSc/KarmaGap-(Copy)?node-id=0-1&node-type=canvas&t=35a2wwsOXYUqahTO-0)

You should look for this section:

Image

### PR DESCRIPTION
**Pull request by heronlancellot**
Closes https://github.com/blockful-io/trustful-gap-deployments/issues/3

### CODE REVIEW

File: components/Pages/Project/Review/CardReviewSummary.tsx [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+   * Calculates the average review score and updates the state with the result.
+   *
+   * @param score - An array of numbers representing individual review scores.
+   * @param grantStoriesLength - The total number of grant stories.
+   */
+  const getAverageReview = (score: number[], grantStoriesLength: number) => {
+    const averageScoreSum = score.reduce(
+      (accumulator, current) => Number(accumulator) + Number(current),
+      0,
+    );
```
Comment (FrancoAguzzi) [inline components/Pages/Project/Review/CardReviewSummary.tsx]:
```suggestion
    const scoresSummed = reviewsScores.reduce(
      (accumulator, current) => Number(accumulator) + Number(current),
      0,
    );
```

File: components/Pages/Project/Review/CardReviewSummary.tsx [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+
+    setAverageScoreReview(
+      Number(averageScoreSum) / 10 ** SCORER_DECIMALS / Number(grantStoriesLength),
```
Comment (FrancoAguzzi) [inline components/Pages/Project/Review/CardReviewSummary.tsx]:
```suggestion
      Number(scoresSummed) / 10 ** SCORER_DECIMALS / Number(reviewsLength),
```

File: components/Pages/Project/Review/CardReviewSummary.tsx [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+  }, [stories, grantIdFromQueryParam]);
+
+  const getTimestampInterval = (timestamps: number[], timestampNow: number) => {
+    if (!timestamps.length) return;
+    if (!stories.length) return;
+
+    const timestampStart = Math.min(...timestamps);
+    const totalInterval = timestampNow - timestampStart;
+
+    const grantStoriesLength = timestamps.length;
```
Comment (FrancoAguzzi) [inline components/Pages/Project/Review/CardReviewSummary.tsx]:
```suggestion
    const numberOfReviews = timestamps.length;
```

File: components/Pages/Project/Review/CardReviewSummary.tsx [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+    const intervalPerReview = totalInterval / grantStoriesLength;
```
Comment (FrancoAguzzi) [inline components/Pages/Project/Review/CardReviewSummary.tsx]:
```suggestion
    const intervalPerReview = totalInterval / numberOfReviews;
```

File: components/Pages/Project/Review/CardReviewSummary.tsx [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+
+  const score = stories?.map((grantStorie: GrantStory) => grantStorie.averageScore) || [];
+
+  /**
+   * Calculates the average review score and updates the state with the result.
+   *
+   * @param score - An array of numbers representing individual review scores.
+   * @param grantStoriesLength - The total number of grant stories.
+   */
+  const getAverageReview = (score: number[], grantStoriesLength: number) => {
```
Comment (FrancoAguzzi) [inline components/Pages/Project/Review/CardReviewSummary.tsx]:
```suggestion
  const getAverageReviewScore = (reviewsScores: number[], reviewsLength: number) => {
```

File: components/Pages/Project/Review/CardReviewSummary.tsx [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+    setAverageScoreReview(
+      Number(averageScoreSum) / 10 ** SCORER_DECIMALS / Number(grantStoriesLength),
+    );
+  };
+
+  const grantIdFromQueryParam = searchParams?.get("grantId");
+
+  useEffect(() => {
+    if (stories) {
+      getAverageReview(score, stories.length);
```
Comment (FrancoAguzzi) [inline components/Pages/Project/Review/CardReviewSummary.tsx]:
```suggestion
      getAverageReviewScore(score, stories.length);
```

File: components/Pages/Project/Review/CardReviewSummary.tsx [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+
+    setTimestampInterval(intervalPerReview);
+    setIntervalMessage(getIntervalMessage(intervalPerReview));
+  };
+
+  const getIntervalMessage = (interval: number): string => {
+    const days = interval / (60 * 60 * 24);
+
+    if (days <= 1) {
+      return `Typically reviewed 1 time ${IntervalMessage.PER_DAY}`;
```
Comment (FrancoAguzzi) [inline components/Pages/Project/Review/CardReviewSummary.tsx]:
```suggestion
      return `Typically reviewed ${the number of reviews done so far} time${number of reviews done so far > 1 ? "s" : ""} ${IntervalMessage.PER_DAY}`;
```

Please put the correct constants inside the ${} I put

File: components/Pages/Project/Review/CardReviewSummary.tsx [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+  };
+
+  const filterScoreReviews = (scoreByEachStarRating: number[], targetScore: number) => {
+    return scoreByEachStarRating.filter(
+      (scoreByEachStarRating) => scoreByEachStarRating === targetScore,
+    ).length;
+  };
+
+  /**
+   * Calculates and sets the star rating data based on the average scores of stories.
```
Comment (FrancoAguzzi) [inline components/Pages/Project/Review/CardReviewSummary.tsx]:
```suggestion
   * Calculates and sets each available score selection percentual based on the total selection each score had 
   * alongside with the total scores selections the grant program had.
```

File: components/Pages/Project/Review/CardReviewSummary.tsx [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+   *
+   * This function maps through the stories, rounds their average scores, and filters them
+   * by each star rating (1 to 5). It then calculates the percentage of each star rating
+   * relative to the total number of stories and sets the star rating data.
```
Comment (FrancoAguzzi) [inline components/Pages/Project/Review/CardReviewSummary.tsx]:
```suggestion
   * relative to the total number of reviews and sets the ratings proportional selections.
```

File: components/Pages/Project/Review/CardReviewSummary.tsx [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+  const filterScoreReviews = (scoreByEachStarRating: number[], targetScore: number) => {
+    return scoreByEachStarRating.filter(
+      (scoreByEachStarRating) => scoreByEachStarRating === targetScore,
+    ).length;
+  };
+
+  /**
+   * Calculates and sets the star rating data based on the average scores of stories.
+   *
+   * This function maps through the stories, rounds their average scores, and filters them
```
Comment (FrancoAguzzi) [inline components/Pages/Project/Review/CardReviewSummary.tsx]:
```suggestion
   * This function maps through the reviews, rounds their average scores, and filters them
```

File: components/Pages/Project/Review/CardReviewSummary.tsx [hunk 11]
```diff
@@ -0,0 +1,279 @@
...
+    } else if (days <= 30) {
+      return `Typically reviewed ${Math.round(30 / days)} times ${IntervalMessage.PER_MONTH}`;
+    } else if (days <= 365) {
+      return `Typically reviewed ${Math.round(365 / days)} times ${IntervalMessage.PER_YEAR}`;
+    } else {
+      return `Typically reviewed each ${Math.round(days)} days`;
+    }
+  };
+
+  const filterScoreReviews = (scoreByEachStarRating: number[], targetScore: number) => {
```
Comment (FrancoAguzzi) [inline components/Pages/Project/Review/CardReviewSummary.tsx]:
```suggestion
  const getNumberOfReviewsPerRating = (allReviews: number[], targetedRating: number) => {
```

File: components/Pages/Project/Review/CardReviewSummary.tsx [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+    return scoreByEachStarRating.filter(
+      (scoreByEachStarRating) => scoreByEachStarRating === targetScore,
+    ).length;
+  };
```
Comment (FrancoAguzzi) [inline components/Pages/Project/Review/CardReviewSummary.tsx]:
```suggestion
    return allReviews.filter(
      (rating) => rating === targetedRating,
    ).length;
  };
```

File: components/Pages/Project/Review/CardReviewSummary.tsx [hunk 13]
```diff
@@ -0,0 +1,279 @@
...
+   *
+   * This function maps through the stories, rounds their average scores, and filters them
+   * by each star rating (1 to 5). It then calculates the percentage of each star rating
+   * relative to the total number of stories and sets the star rating data.
+   *
+   */
+  const getScoreRatingFilteredReviews = () => {
+    const scoreByEachStarRating = stories.map((story: GrantStory) =>
+      Math.round(Number(story.averageScore) / 10 ** SCORER_DECIMALS),
+    );
```
Comment (FrancoAguzzi) [inline components/Pages/Project/Review/CardReviewSummary.tsx]:
```suggestion
    const scoresOfReviews = stories.map((story: GrantStory) =>
      Math.round(Number(story.averageScore) / 10 ** SCORER_DECIMALS),
    );
```

File: components/Pages/Project/Review/CardReviewSummary.tsx [hunk 14]
```diff
@@ -0,0 +1,279 @@
...
+
+    const starDataFiltered = {
+      1: filterScoreReviews(scoreByEachStarRating, 1),
+      2: filterScoreReviews(scoreByEachStarRating, 2),
+      3: filterScoreReviews(scoreByEachStarRating, 3),
+      4: filterScoreReviews(scoreByEachStarRating, 4),
+      5: filterScoreReviews(scoreByEachStarRating, 5),
```
Comment (FrancoAguzzi) [inline components/Pages/Project/Review/CardReviewSummary.tsx]:
```suggestion
      1: getNumberOfReviewsPerRating(scoresOfReviews, 1),
      2: getNumberOfReviewsPerRating(scoresOfReviews, 2),
      3: getNumberOfReviewsPerRating(scoresOfReviews, 3),
      4: getNumberOfReviewsPerRating(scoresOfReviews, 4),
      5: getNumberOfReviewsPerRating(scoresOfReviews, 5),
```

File: components/Pages/Project/Review/CardReviewSummary.tsx [hunk 15]
```diff
@@ -0,0 +1,279 @@
...
+   * by each star rating (1 to 5). It then calculates the percentage of each star rating
+   * relative to the total number of stories and sets the star rating data.
+   *
+   */
+  const getScoreRatingFilteredReviews = () => {
+    const scoreByEachStarRating = stories.map((story: GrantStory) =>
+      Math.round(Number(story.averageScore) / 10 ** SCORER_DECIMALS),
+    );
+
+    const starDataFiltered = {
```
Comment (FrancoAguzzi) [inline components/Pages/Project/Review/CardReviewSummary.tsx]:
```suggestion
    const reviewsFilteredByScore = {
```

File: components/Pages/Project/Review/CardReviewSummary.tsx [hunk 16]
```diff
@@ -0,0 +1,279 @@
...
+      1: filterScoreReviews(scoreByEachStarRating, 1),
+      2: filterScoreReviews(scoreByEachStarRating, 2),
+      3: filterScoreReviews(scoreByEachStarRating, 3),
+      4: filterScoreReviews(scoreByEachStarRating, 4),
+      5: filterScoreReviews(scoreByEachStarRating, 5),
+    };
+
+    const starsRatingData: StarRatingDataProps[] = Object.entries(starDataFiltered).map(
```
Comment (FrancoAguzzi) [inline components/Pages/Project/Review/CardReviewSummary.tsx]:
```suggestion
    const reviewsWithPercentualRelevance: StarRatingDataProps[] = Object.entries(reviewsFilteredByScore).map(
```

File: components/Pages/Project/Review/CardReviewSummary.tsx [hunk 17]
```diff
@@ -0,0 +1,279 @@
...
+      ([stars, count]) => {
+        const percentage = (count / stories.length) * 100;
+        return {
+          stars: Number(stars),
+          count: count,
+          percentage: percentage,
+        };
+      },
+    );
+    setStarRatingFiltered(starsRatingData);
```
Comment (FrancoAguzzi) [inline components/Pages/Project/Review/CardReviewSummary.tsx]:
```suggestion
    setStarRatingFiltered(reviewsWithPercentualRelevance);
```

File: components/Pages/Project/Review/CardReviewSummary.tsx [hunk 18]
```diff
@@ -0,0 +1,279 @@
...
+export const CardReviewSummary = () => {
+  const project = useProjectStore((state: any) => state.project);
+  const setIsOpenReview = useReviewStore((state: any) => state.setIsOpenReview);
+  const setActiveBadges = useReviewStore((state: any) => state.setActiveBadges);
+  const setActiveBadgeIds = useReviewStore((state: any) => state.setActiveBadgeIds);
+  const stories = useReviewStore((state: any) => state.stories);
+  const [timestampInterval, setTimestampInterval] = useState<number | null>(null);
+  const [averageScoreReview, setAverageScoreReview] = useState<number | null>(null);
+  const [intervalMessage, setIntervalMessage] = useState<string | undefined>(undefined);
+  const [starRatingFiltered, setStarRatingFiltered] = useState<StarRatingDataProps[]>([]);
```
Comment (FrancoAguzzi) [inline components/Pages/Project/Review/CardReviewSummary.tsx]: I'd call it `reviewsScoreWithPercentage`

File: components/Pages/Project/Review/ProgressBar.tsx [hunk 19]
```diff
@@ -0,0 +1,23 @@
+interface ProgressBarProps {
+  currentStep: number;
+  numberOfItems: number;
+}
+
+export const ProgressBar = ({ currentStep, numberOfItems = 100 }: ProgressBarProps) => {
```
Comment (FrancoAguzzi) [inline components/Pages/Project/Review/ProgressBar.tsx]: It is not clear to me what a prop called number of items has in common to a ProgressBar, can it be renamed?

**Review by FrancoAguzzi — CHANGES_REQUESTED** [review]

File: components/Pages/Project/Review/CardReviewSummary.tsx (hunk 18 above)
Comment (heronlancellot) [reply components/Pages/Project/Review/CardReviewSummary.tsx]: what state ?

File: components/Pages/Project/Review/ProgressBar.tsx [hunk 20]
```diff
@@ -1,19 +1,20 @@
-interface ProgressBarProps {
-  currentStep: number;
-  numberOfItems: number;
-}
+/**
+ * ProgressBar component displays a progress bar indicating the completion percentage.
+ *
```
Comment (FrancoAguzzi) [inline components/Pages/Project/Review/ProgressBar.tsx]: Please remove the additional *!

**Review by FrancoAguzzi — APPROVED** [review]

File: components/Pages/Project/Review/CardReviewSummary.tsx (hunk 18 above)
Comment (FrancoAguzzi) [reply components/Pages/Project/Review/CardReviewSummary.tsx]: The one tagged! 😆

**Review by FrancoAguzzi — APPROVED** [review]

**Pull request merged**